email.go

```diff
@@ -98,9 +105,115 @@ func (c emailConfig) SendValidation(domainInfo *db.DomainData, token string) err
 }
 
 func (c emailConfig) sendEmail(subject string, body string, address string) error {
+	blacklisted, err := c.database.IsBlacklistedEmail(address)
+	if err != nil {
+		return err
+	}
+	if blacklisted {
+		return fmt.Errorf("address %s is blacklisted", address)
+	}
 	message := fmt.Sprintf("From: %s\nTo: %s\nSubject: %s\n\n%s",
 		c.sender, address, subject, body)
 	return smtp.SendMail(fmt.Sprintf("%s:%s", c.submissionHostname, c.port),
 		c.auth,
 		c.sender, []string{address}, []byte(message))
 }
+
+type blacklistRequest struct {
+	reason     string
+	timestamp  string
+	recipients []struct {
+		EmailAddress string `json:"emailAddress"`
+	}
+}
+
+// UnmarshallJSON wrangles the JSON posted by AWS SNS into something easier to access
+// and generalized across notification types.
+func (r *blacklistRequest) UnmarshalJSON(b []byte) error {
+	var wrapper struct {
+		Type      string
+		MessageID string
+		Message   string
+		Timestamp string
+	}
+	if err := json.Unmarshal(b, &wrapper); err != nil {
+		return fmt.Errorf("failed to load notification wrapper: %v", err)
+	}
+
+	var msg struct {
+		NotificationType string `json:"notificationType"`
+		Complaint        struct {
+			ComplainedRecipients []struct {
+				EmailAddress string `json:"emailAddress"`
+			} `json:"complainedRecipients"`
+			Timestamp string `json:"timestamp"`
+		} `json:"complaint"`
+		Bounce struct {
+			BouncedRecipients []struct {
+				EmailAddress string `json:"emailAddress"`
+			} `json:"bouncedRecipients"`
+			Timestamp string `json:"timestamp"`
+		} `json:"bounce"`
+	}
+	if err := json.Unmarshal([]byte(wrapper.Message), &msg); err != nil {
+		return fmt.Errorf("failed to load notification message: %v", err)
+	}
+
+	switch msg.NotificationType {
+	case "Complaint":
+		*r = blacklistRequest{
+			reason:     msg.NotificationType,
+			timestamp:  msg.Complaint.Timestamp,
+			recipients: msg.Complaint.ComplainedRecipients,
+		}
+	case "Bounce":
+		*r = blacklistRequest{
+			reason:     msg.NotificationType,
+			timestamp:  msg.Bounce.Timestamp,
+			recipients: msg.Bounce.BouncedRecipients,
+		}
+	default:
+		return fmt.Errorf("SES notification did not match expected format")
+	}
+
+	return nil
+}
+
+// handleSESNotification handles AWS SES bounces and complaints submitted to a webhook
+// via AWS SNS (Simple Notification Service).
+// The SNS webwook is configured to include a secret API key stored in the environment.
+func handleSESNotification(database db.Database) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		keyParam := r.URL.Query()["amazon_authorize_key"]
+		if len(keyParam) == 0 || keyParam[0] != os.Getenv("AMAZON_AUTHORIZE_KEY") {
```

Comment:
do we 401 all requests to `/sns` if this var isn't set? We might want to do something explicitly; could be as simple/unobtrusive as dropping a message in the `log` just to let the operator know something's up, or returning 5** to let the client know it's not their fault.

Reply:
also, a follow-up question about this line: is `keyParam` a string? Should it be `keyParam != os.Getenv(...)`?